rx_signal_processing/decimate.cu

```diff
@@ -1,221 +1,693 @@
-
-    auto filter_tap_num = threadIdx.x;
-    auto num_filter_taps = blockDim.x;
-    cuComplex total_sum = data[tap_offset];
 
+#include <cuComplex.h> //cuComplex type and all cuCmulf/cuCaddf functions.
+#include <iostream>
+#include <stdint.h>
+#include "decimate.hpp"
+//This keeps the contained functions local to this file.
+namespace {
+  /**
+   * @brief      Creates a new set of grid dimensions for a bandpass decimate CUDA kernel.
+   *
+   * @param[in]  num_samples   Number of input samples.
+   * @param[in]  dm_rate       Decimation rate.
+   * @param[in]  num_antennas  Number of antennas for which there are samples.
+   *
+   * @return     New grid dimensions for the kernel.
+   */
+  dim3 create_bandpass_grid(uint32_t num_samples, uint32_t dm_rate, uint32_t num_antennas)
+  {
+    auto num_blocks_x = num_samples/dm_rate;
+    auto num_blocks_y = num_antennas;
+    auto num_blocks_z = 1;
+    std::cout << "    Grid size: " << num_blocks_x << " x " << num_blocks_y << " x "
+      << num_blocks_z << std::endl;
+    dim3 dimGrid(num_blocks_x,num_blocks_y,num_blocks_z);
 
-    if ((num_filter_taps >= 512) && (filter_tap_num < 256))
-    {
-        total_sum = cuCaddf(total_sum,data[tap_offset  + 256]);
-        data[tap_offset] = total_sum; // REVIEW #25 Is it necessary for speed to have two '=' statements on one line? it took a while to see the second one, therefore more confusing. split into two lines
-    }                                       // REPLY okay
-
-    __syncthreads();
+    return dimGrid;
+  }
+
+  /**
+   * @brief      Creates a new set of block dimensions for a bandpass decimate CUDA kernel.
+   *
+   * @param[in]  num_taps_per_filter  Number of taps per filter.
+   * @param[in]  num_freqs            Number of receive frequencies.
+   *
+   * @return     New block dimensions for the kernel.
+   */
+  dim3 create_bandpass_block(uint32_t num_taps_per_filter, uint32_t num_freqs)
+  {
+    auto num_threads_x = num_taps_per_filter;
+    auto num_threads_y = num_freqs;
+    auto num_threads_z = 1;
+    std::cout << "    Block size: " << num_threads_x << " x " << num_threads_y << " x "
+      << num_threads_z << std::endl;
+    dim3 dimBlock(num_threads_x,num_threads_y,num_threads_z);
 
-    if ((num_filter_taps >= 256) && (filter_tap_num < 128))
-    {
-        total_sum = cuCaddf(total_sum, data[tap_offset + 128]);
-        data[tap_offset] = total_sum;
-    }
+    return dimBlock;
+  }
+
+  /**
+   * @brief      Creates a new set of grid dimensions for a lowpass decimate CUDA kernel.
+   *
+   * @param[in]  num_samples   Number of input samples in a frequency dataset.
+   * @param[in]  dm_rate       Decimation rate.
+   * @param[in]  num_antennas  Number of antennas for which there are samples.
+   * @param[in]  num_freqs     Number of receive frequencies.
+   *
+   * @return     New grid dimensions for the kernel.
+   */
+  dim3 create_lowpass_grid(uint32_t num_samples, uint32_t dm_rate, uint32_t num_antennas,
```

Comment:
Is lowpass_decimate2048 ever going to be used because of your setup where num_taps = number of threads per block? Are we going to have filters over 1024 taps? I like this lowpass setup of grids and blocks where blockDim.y and blockDim.z = 1. Just wondering why it can't be used for bandpass as well to reduce (or eliminate the need for) the use of bandpass_decimate2048. The only difference in the bandpass would be feeding different filters to each frequency in the gridDim.z direction instead of different datasets as you're doing here.